superduperdb/components/template.py

```diff
@@ -0,0 +1,73 @@
+import dataclasses as dc
+import typing as t
+
+from superduperdb.base.document import Document
+from superduperdb.components.datatype import pickle_serializer, _BaseEncodable
+from superduperdb.misc.annotations import merge_docstrings
+from superduperdb.base.variables import Variable
+
+from .component import Component, ensure_initialized
+
+if t.TYPE_CHECKING:
+    from superduperdb import DataType
+
+
+@merge_docstrings
+@dc.dataclass(kw_only=True)
+class Template(Component):
+    """
+    Application template component.
+
+    :param template: Template.
+    :param info: Info.
+    """
+
+    type_id: t.ClassVar[str] = 'template'
+
+    component: t.Union[Component, t.Dict]
+    info: t.Dict
+
+    def __post_init__(self, db, artifacts):
+        self._variables = []
+        if isinstance(self.component, Component):
+            self._variables  = self.component.variables
+            self.component = self.component.dict().encode()
+        return super().__post_init__(db, artifacts)
+
+
+    @ensure_initialized
+    def __call__(self, **kwargs):
+        """Method to create component from the given template and `kwargs`."""
+        if self.info:
+            assert set(kwargs.keys()) == (set(self.info.keys())), 'Invalid variables'
+        if isinstance(self.component, dict):
+            t = Document.decode(self.component, db=self.db).unpack()
+        else:
+            t = self.component
+
+        t = t.set_variables(db=self.db, **kwargs)
+        t.init(db=self.db)
+        return t
+
+    @property
+    def variables(self):
+        return self._variables
+    
+
+    def dict(self):
+        from superduperdb import Document
+        r = super().dict().unpack()
+
+        leaves = r.get('_leaves', {})
```

Comment:
Move this logic to `.encode`.

Reply:
as discussed it is required in `dict` only. since data layer._add runs .dict().encode()